cmd/gaia/app/export.go

```diff
@@ -0,0 +1,119 @@
+package app
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/auth"
+	distr "github.com/cosmos/cosmos-sdk/x/distribution"
+	"github.com/cosmos/cosmos-sdk/x/gov"
+	"github.com/cosmos/cosmos-sdk/x/mint"
+	"github.com/cosmos/cosmos-sdk/x/slashing"
+	stake "github.com/cosmos/cosmos-sdk/x/stake"
+	abci "github.com/tendermint/tendermint/abci/types"
+	tmtypes "github.com/tendermint/tendermint/types"
+)
+
+// export the state of gaia for a genesis file
+func (app *GaiaApp) ExportAppStateAndValidators(forZeroHeight bool) (appState json.RawMessage, validators []tmtypes.GenesisValidator, err error) {
+	// as if they could withdraw from the start of the next block
+	ctx := app.NewContext(true, abci.Header{Height: app.LastBlockHeight()})
+
+	// prepare for fresh start at zero height
+	if forZeroHeight {
+
+		/* Handle fee distribution state. */
+
+		// withdraw all delegator & validator rewards
+		app.distrKeeper.IterateValidatorDistInfos(ctx, func(_ int64, valInfo distr.ValidatorDistInfo) (stop bool) {
+			err := app.distrKeeper.WithdrawValidatorRewardsAll(ctx, valInfo.OperatorAddr)
+			if err != nil {
+				panic(err)
+			}
+			return false
+		})
+		app.distrKeeper.IterateDelegationDistInfos(ctx, func(_ int64, distInfo distr.DelegationDistInfo) (stop bool) {
+			app.distrKeeper.WithdrawDelegationRewardsAll(ctx, distInfo.DelegatorAddr)
+			return false
+		})
+
+		// delete all distribution infos
+		app.distrKeeper.RemoveValidatorDistInfos(ctx)
+		app.distrKeeper.RemoveDelegationDistInfos(ctx)
```

Comment:
We call `OnDelegationCreated` in the stake genesis handler, which will recreate the dist infos, so I thought it was cleaner to delete them - do you think we should do otherwise (and maybe change the stake handler too)?

Reply:
We call OnDelegationCreated? - yeah the delegation is more re-created - at that point hmm

Reply:
Okay - I think for the purposes of this PR it's okay to just leave this here, HOWEVER - we should absolutely leave a comment here explaining that the dist-info's are recreated on next genesis as the OnDelegationCreated method is called. 

In the future, I think we _should_ restructure the genesis logic a bit and include a genesis parameter in staking which specifies whether this is a new chain or a chain-restart, the genesis if in genesis restart mode, then we ought to _not_ delete the `DistInfo`s and export them instead. This will be especially important for the "restart tendermint at a given height" ultimate solution which doesn't require withdrawing any distribution funds.  

reference issue just opened
https://github.com/cosmos/cosmos-sdk/issues/2862

REMAINING ACTION ITEM: just need to update the comment here for now I think

Reply:
Agreed in general, commented on the issue, export.go comment updated.